### PR TITLE
Make check.inst() perform type-narrowing

### PR DESCRIPTION
**Pull request by smackesey**
## Summary & Motivation

Make `check.inst` derive the returned type from the passed-in type to runtime-check, thereby performing type-narrowing. This did not used to be possible but improvements to type variable resolution in later versions of pyright support it.

## How I Tested These Changes

Converted a few instances that were both casting and `check.inst`-ing to use `check.narrow`.

Upstack I convert more `cast` instances.

### CODE REVIEW

**Comment by smackesey** [comment]
* **#20687** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20687?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @smackesey and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by alangenfeld** [comment]
>  often the check.inst behavior is preferable

Something still feels wrong to me about the static type checking ignoring what the runtime validation is doing. Was this the opinion you had going in to making this change or did you start by making `inst` narrow and the reality observed convinced you of this position? 

> check.narrow(x, Foo) # => x is now Foo[Any]

Does `check.narrow(x, Foo[Bar])` work or is there a limitation preventing parameterized stuff from flowing through?

**Comment by schrockn** [comment]
Would renaming `check.inst` to `check.typed_hinted_inst` or something (ideally more concise) make its purpose more clear? Effectively all it is going is making sure a passed in object matches its signature. I think this is why `inst_param` feels better since it is just "make sure the param type hint isn't lying", which make more sense.

**Comment by smackesey** [comment]
Having thought about this a little more, I now think it makes sense to just alter `check.inst` because there is no case in which we need to capture the return value in the non-narrowing case.

Right now, because `check.inst` (a) never alters the passed-in value; (b) never alters the passed-in type annotation, there is no point in ever doing this:

```
x = check.inst(x, ...)
# or
x = check.inst_param(x, ...)
```

Instead it should just be called without reassignment. So if we make `check.inst` perform narrowing, then we just do a non-reassignment when you want to preserve type info and do a reassignment when we want to narrow.

File: python_modules/dagster/dagster/_check/__init__.py [hunk 1]
```diff
@@ -613,9 +613,22 @@ def opt_int_elem(
...
+#     inst(foo, Foo[Bar])  # illegal, can't pass parameterized types to inst()
+#     inst(foo, Foo)  # type is Foo; because we couldn't pass parameterized type, we lost info
+#
+# In contrast, we don't lose type information when we pass the type of the checked argument through
+# without modification.
+#
+# Because of this trade-off, it makes sense for inst() to perform narrowing but not inst_param().
+# With inst(), we rarely have rich type information at the callsite (if we did we wouldn't be
+# calling inst()). With inst_param(), on the other hand, we should always have rich type information
+# at the callsite from the type annotation, so we should never need to narrow.
```
Comment (schrockn) [inline python_modules/dagster/dagster/_check/__init__.py:631]: Should we just rename `inst` to `narrow`? I think it would make all this more intuitive,

Comment (smackesey) [reply python_modules/dagster/dagster/_check/__init__.py:631]: >Should we just rename inst to narrow? I think it would make all this more intuitive,

I don't think so, because (a) it would disrupt the existing naming scheme of "param" being a modifier to an existing method; (b) I think this will almost entirely "just work" as people want, so they won't need to dig into the details.

Comment (schrockn) [reply python_modules/dagster/dagster/_check/__init__.py:631]: I buy it 👍🏻

File: python_modules/dagster/dagster/_config/field.py [hunk 2]
```diff
@@ -305,7 +305,7 @@ def __init__(
                         "into of a config enum type {name}. You must pass in the underlying "
                         "string represention as the default value. One of {value_set}."
                     ).format(
-                        value_set=[ev.config_value for ev in self.config_type.enum_values],
+                        value_set=[ev.config_value for ev in self.config_type.enum_values],  # type: ignore
```
Comment (schrockn) [inline python_modules/dagster/dagster/_config/field.py:308]: unrelated?

**Review by schrockn — APPROVED** [review]

File: python_modules/dagster/dagster/_config/field.py (hunk 2 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_config/field.py:308]: not unrelated, since `inst()` performs narrowing, we go from `Unknown/Any` to `ConfigType` above, resulting in the need for this `type: ignore` (since only a subclass of `ConfigType` have `enum_values`)
